File: cm-backend/src/admin_backend/routers/v1/onboarding.py
"""Client-onboarding wizard section + state endpoints.

All routes are tenant-scoped under ``/tenants/{tenant_id}`` and gated
``ADMIN.TENANTS.CONFIGURE.GLOBAL`` with ``audience="PLATFORM"`` (staff
onboarding, D-12). Section reads/writes and the onboarding-state
resource live here; ``complete-onboarding`` stays on the tenants router
(it is a tenant lifecycle transition).

Sits behind the same ``/tenants`` prefix as ``tenants_router``; FastAPI
composes both routers under that prefix without conflict (distinct
paths).
"""
from __future__ import annotations

from typing import Any
from uuid import UUID

from fastapi import APIRouter, Depends, Request
from sqlalchemy.ext.asyncio import AsyncSession

from admin_backend.auth.context import AuthContext
from admin_backend.auth.permissions import require
from admin_backend.dependencies import (
    get_auth_context,
    get_tenant_session_dep,
)
from admin_backend.errors import (
    OnboardingSectionNotFoundError,
    TenantNotFoundError,
)
from admin_backend.models.permission import (
    PermissionAction,
    PermissionResource,
    PermissionScope,
)
from admin_backend.models.tenant_module_access import ModuleCode
from admin_backend.repositories.onboarding import OnboardingRepo
from admin_backend.schemas.onboarding import (
    BillingProfileRead,
    BillingProfileUpsertRequest,
    ContactItem,
    ContactsRead,
    ContactsReplaceRequest,
    LegalProfileRead,
    LegalProfileUpsertRequest,
    OnboardingPatchRequest,
    OnboardingStateResponse,
    TaxRegistrationItem,
    TaxRegistrationsRead,
    TaxRegistrationsReplaceRequest,
)


router = APIRouter(prefix="/tenants", tags=["onboarding"])

_repo = OnboardingRepo()


def _gate() -> Any:
    """One gate for every onboarding section/state route:
    ``ADMIN.TENANTS.CONFIGURE.GLOBAL``, PLATFORM audience."""
    return require(
        ModuleCode.ADMIN,
        PermissionResource.TENANTS,
        PermissionAction.CONFIGURE,
        PermissionScope.GLOBAL,
        audience="PLATFORM",
    )


async def _require_tenant_visible(
    repo: OnboardingRepo, session: AsyncSession, tenant_id: UUID
) -> None:
    if await repo.tenant_name_or_none(session, tenant_id) is None:
        raise TenantNotFoundError(
            f"Tenant {tenant_id} not visible to this session",
            tenant_id=str(tenant_id),
        )


# ---------------------------------------------------------------------------
# Legal profile (1:1)
# ---------------------------------------------------------------------------


@router.get(
    "/{tenant_id}/legal-profile", response_model=LegalProfileRead
)
async def get_legal_profile(
    tenant_id: UUID,
    _: None = Depends(_gate()),
    session: AsyncSession = Depends(get_tenant_session_dep),
) -> Any:
    await _require_tenant_visible(_repo, session, tenant_id)
    row = await _repo.get_legal_profile(session, tenant_id)
    if row is None:
        raise OnboardingSectionNotFoundError(section="Legal profile")
    return LegalProfileRead.model_validate(row)


@router.put(
    "/{tenant_id}/legal-profile", response_model=LegalProfileRead
)
async def put_legal_profile(
    tenant_id: UUID,
    body: LegalProfileUpsertRequest,
    request: Request,
    _: None = Depends(_gate()),
    auth: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_tenant_session_dep),
) -> Any:
    row = await _repo.upsert_legal_profile(
        session,
        tenant_id,
        legal_entity_name=body.legal_entity_name,
        entity_type=body.entity_type,
        registration_number=body.registration_number,
        incorporation_date=body.incorporation_date,
        registered_address=body.registered_address,
        actor_user_id=auth.user_id,
        auth=auth,
        request_id=request.state.request_id,
    )
    return LegalProfileRead.model_validate(row)


# ---------------------------------------------------------------------------
# Tax registrations (1:N)
# ---------------------------------------------------------------------------


@router.get(
    "/{tenant_id}/tax-registrations", response_model=TaxRegistrationsRead
)
async def get_tax_registrations(
    tenant_id: UUID,
    _: None = Depends(_gate()),
    session: AsyncSession = Depends(get_tenant_session_dep),
) -> Any:
    await _require_tenant_visible(_repo, session, tenant_id)
    rows = await _repo.list_tax_registrations(session, tenant_id)
    return TaxRegistrationsRead(
        items=[TaxRegistrationItem.model_validate(r) for r in rows]
    )


@router.put(
    "/{tenant_id}/tax-registrations", response_model=TaxRegistrationsRead
)
async def put_tax_registrations(
    tenant_id: UUID,
    body: TaxRegistrationsReplaceRequest,
    request: Request,
    _: None = Depends(_gate()),
    auth: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_tenant_session_dep),
) -> Any:
    rows = await _repo.replace_tax_registrations(
        session,
        tenant_id,
        items=[it.model_dump() for it in body.items],
        actor_user_id=auth.user_id,
        auth=auth,
        request_id=request.state.request_id,
    )
    return TaxRegistrationsRead(
        items=[TaxRegistrationItem.model_validate(r) for r in rows]
    )


# ---------------------------------------------------------------------------
# Billing profile (1:1)
# ---------------------------------------------------------------------------


@router.get(
    "/{tenant_id}/billing-profile", response_model=BillingProfileRead
)
async def get_billing_profile(
    tenant_id: UUID,
    _: None = Depends(_gate()),
    session: AsyncSession = Depends(get_tenant_session_dep),
) -> Any:
    await _require_tenant_visible(_repo, session, tenant_id)
    row = await _repo.get_billing_profile(session, tenant_id)
    if row is None:
        raise OnboardingSectionNotFoundError(section="Billing profile")
    return BillingProfileRead.model_validate(row)


@router.put(
    "/{tenant_id}/billing-profile", response_model=BillingProfileRead
)
async def put_billing_profile(
    tenant_id: UUID,
    body: BillingProfileUpsertRequest,
    request: Request,
    _: None = Depends(_gate()),
    auth: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_tenant_session_dep),
) -> Any:
    row = await _repo.upsert_billing_profile(
        session,
        tenant_id,
        payment_terms=body.payment_terms,
        currency=body.currency,
        billing_email=body.billing_email,
        billing_contact_name=body.billing_contact_name,
        billing_address=body.billing_address,
        actor_user_id=auth.user_id,
        auth=auth,
        request_id=request.state.request_id,
    )
    return BillingProfileRead.model_validate(row)


# ---------------------------------------------------------------------------
# Contacts (1:N)
# ---------------------------------------------------------------------------


@router.get("/{tenant_id}/contacts", response_model=ContactsRead)
async def get_contacts(
    tenant_id: UUID,
    _: None = Depends(_gate()),
    session: AsyncSession = Depends(get_tenant_session_dep),
) -> Any:
    await _require_tenant_visible(_repo, session, tenant_id)
    rows = await _repo.list_contacts(session, tenant_id)
    return ContactsRead(items=[ContactItem.model_validate(r) for r in rows])


@router.put("/{tenant_id}/contacts", response_model=ContactsRead)
async def put_contacts(
    tenant_id: UUID,
    body: ContactsReplaceRequest,
    request: Request,
    _: None = Depends(_gate()),
    auth: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_tenant_session_dep),
) -> Any:
    rows = await _repo.replace_contacts(
        session,
        tenant_id,
        items=[it.model_dump() for it in body.items],
        actor_user_id=auth.user_id,
        auth=auth,
        request_id=request.state.request_id,
    )
    return ContactsRead(items=[ContactItem.model_validate(r) for r in rows])


# ---------------------------------------------------------------------------
# Onboarding state resource
# ---------------------------------------------------------------------------


@router.get("/{tenant_id}/onboarding", response_model=OnboardingStateResponse)
async def get_onboarding(
    tenant_id: UUID,
    _: None = Depends(_gate()),
    session: AsyncSession = Depends(get_tenant_session_dep),
) -> Any:
    state = await _repo.get_onboarding_state(session, tenant_id)
    if state is None:
        raise TenantNotFoundError(
            f"Tenant {tenant_id} not visible to this session",
            tenant_id=str(tenant_id),
        )
    return OnboardingStateResponse.model_validate(state)


@router.patch(
    "/{tenant_id}/onboarding", response_model=OnboardingStateResponse
)
async def patch_onboarding(
    tenant_id: UUID,
    body: OnboardingPatchRequest,
    request: Request,
    _: None = Depends(_gate()),
    auth: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_tenant_session_dep),
) -> Any:
    sent = body.model_dump(exclude_unset=True)
    state = await _repo.patch_onboarding(
        session,
        tenant_id,
        current_step=body.current_step,
        current_step_set="current_step" in sent,
        section_status=body.section_status,
        section_status_set="section_status" in sent,
        actor_user_id=auth.user_id,
        auth=auth,
        request_id=request.state.request_id,
    )
    return OnboardingStateResponse.model_validate(state)

File: cm-frontend/lib/api/onboarding.ts
import { ApiError, apiFetch } from "./client";
import type {
  BillingProfileRead,
  BillingProfileUpsertRequest,
  ContactsRead,
  ContactsReplaceRequest,
  LegalProfileRead,
  LegalProfileUpsertRequest,
  OnboardingPatchRequest,
  OnboardingStateResponse,
  TaxRegistrationsRead,
  TaxRegistrationsReplaceRequest,
} from "@/types/api";

// Client-onboarding wizard data layer over the backend section + state
// endpoints. All tenant-scoped under /tenants/{tenant_id}; every route
// is PLATFORM-audience-gated server-side.
//
// The 1:1 section GETs (legal-profile, billing-profile) return 404
// SECTION_NOT_FOUND when the section has not been saved yet (distinct from
// TENANT_NOT_FOUND). The wizard treats "not saved yet" as an empty form,
// so those getters translate that specific 404 to null and let any other
// error propagate.
//
// Writes carry a per-call Idempotency-Key, matching the tenants write
// convention (harmless for the naturally-idempotent upsert / full-replace
// section writes; keeps the write surface uniform).

function writeHeaders(): Record<string, string> {
  return { "Idempotency-Key": crypto.randomUUID() };
}

async function getOrNull<T>(path: string): Promise<T | null> {
  try {
    return await apiFetch<T>(path);
  } catch (err) {
    if (err instanceof ApiError && err.code === "SECTION_NOT_FOUND") {
      return null;
    }
    throw err;
  }
}

export const onboardingApi = {
  // --- Legal profile (1:1) ---
  getLegalProfile: (tenantId: string) =>
    getOrNull<LegalProfileRead>(
      `/api/v1/tenants/${tenantId}/legal-profile`,
    ),
  putLegalProfile: (tenantId: string, body: LegalProfileUpsertRequest) =>
    apiFetch<LegalProfileRead>(
      `/api/v1/tenants/${tenantId}/legal-profile`,
      { method: "PUT", body: JSON.stringify(body), headers: writeHeaders() },
    ),

  // --- Tax registrations (1:N full-replace) ---
  getTaxRegistrations: (tenantId: string) =>
    apiFetch<TaxRegistrationsRead>(
      `/api/v1/tenants/${tenantId}/tax-registrations`,
    ),
  putTaxRegistrations: (
    tenantId: string,
    body: TaxRegistrationsReplaceRequest,
  ) =>
    apiFetch<TaxRegistrationsRead>(
      `/api/v1/tenants/${tenantId}/tax-registrations`,
      { method: "PUT", body: JSON.stringify(body), headers: writeHeaders() },
    ),

  // --- Billing profile (1:1) ---
  getBillingProfile: (tenantId: string) =>
    getOrNull<BillingProfileRead>(
      `/api/v1/tenants/${tenantId}/billing-profile`,
    ),
  putBillingProfile: (tenantId: string, body: BillingProfileUpsertRequest) =>
    apiFetch<BillingProfileRead>(
      `/api/v1/tenants/${tenantId}/billing-profile`,
      { method: "PUT", body: JSON.stringify(body), headers: writeHeaders() },
    ),

  // --- Contacts (1:N full-replace) ---
  getContacts: (tenantId: string) =>
    apiFetch<ContactsRead>(`/api/v1/tenants/${tenantId}/contacts`),
  putContacts: (tenantId: string, body: ContactsReplaceRequest) =>
    apiFetch<ContactsRead>(`/api/v1/tenants/${tenantId}/contacts`, {
      method: "PUT",
      body: JSON.stringify(body),
      headers: writeHeaders(),
    }),

  // --- Onboarding state (resume truth) ---
  getState: (tenantId: string) =>
    apiFetch<OnboardingStateResponse>(
      `/api/v1/tenants/${tenantId}/onboarding`,
    ),
  patchState: (tenantId: string, body: OnboardingPatchRequest) =>
    apiFetch<OnboardingStateResponse>(
      `/api/v1/tenants/${tenantId}/onboarding`,
      { method: "PATCH", body: JSON.stringify(body), headers: writeHeaders() },
    ),
};
